static/js/code-highlighter.js

```diff
@@ -8,6 +8,232 @@
+      splitPoint = -1;
+    } else {
+      splitPoint = -2;
+    }
+
+    const nsPieces = pieces.slice(0, splitPoint);
+    const fullPieces = pieces.slice(splitPoint);
+
+    const nsTransformed = nsPieces.map(piece => {
+      return piece[0];
+    });
+
+    if (nsTransformed.length === 0) {
+      return fullPieces.join("::");
+    }
+    return nsTransformed.join(":") + ":" + fullPieces.join("::");
+  }
+
+  /**
+   * Given an element corresponding to a source line, figure out the most
+   * appropriate pretty symbol in the line.  This is currently done using a
+   * heuristic, but this should ideally be handled by either:
+   * - Directly annotating the DOM with the symbol element that is inducing
+   *   the nesting.
+   * - Switching from the data-i scheme to using a symbols dictionary and
+   *   ensuring the symbol data similarly identifies any nesting associated with
+   *   the symbol.
+   *
+   * The current heuristic logic is:
+   * - Pick the last observed symbol (by having a "data-i" attribute) preceding
+   *   a `(`.
+   */
+  _findBestPrettySymbolInSourceLineElem(elem) {
+    let bestPretty = null;
+    if (!elem) {
+      return bestPretty;
+    }
+
+    const symElems = elem.querySelectorAll("[data-i]");
+    scan: for (const symElem of symElems) {
+      // Check if any of the preceding nodes had a "(" in them.  If they did,
+      // this symbol is irrelevant and we should break out of the outer "scan"
+      // loop.
+      //
+      // Okay, and now we're gaining one more hacky heuristic to deal with the
+      // situation "class Foo : public DontCare, public AlsoDontCare {".  If we
+      // see "class" and " : ", we also bail.  The complication here is that we
+      // do absolutely want to pick "Bar" in "Foo::Bar()", so we can't just bail
+      // when we see a colon anywhere.
+      let sawClass = false;
+      let sawColon = false;
+      for (let prevNode = symElem.previousSibling;
+           prevNode;
+           prevNode = prevNode.previousSibling) {
+        // A "(" always means stop immediately.
+        if (prevNode.textContent.includes("(")) {
+          break scan;
+        }
+
+        // The compound class check.
+        if (prevNode.textContent.includes("class")) {
+          sawClass = true;
+        }
+        if (prevNode.textContent.includes(" : ")) {
+          sawColon = true;
+        }
+        if (sawClass && sawColon) {
+          break scan;
+        }
+      }
+
+      // Extract the most appropriate pretty data from the searches.
+      // Specifically, we are looking for "pretty" text in the searches that
+      // contains the textContent from the semantic token.  We do this to
+      // compensate for the implicitly invoked field constructors which
+      // currently end up coalesced into the constructor's symbol/point.
+      const visibleToken = symElem.textContent;
+      const data = window.ANALYSIS_DATA[symElem.getAttribute("data-i")];
```

Comment:
Are there any semantic differences between reading the attribute and reading `.dataset.i`? Just wondering, since I generally prefer the latter as it fails sooner/better.

Reply:
Honestly, I had the same question but was cargo-culting from [code that already was using getAttribute](https://github.com/mozsearch/mozsearch/blob/989908c990abeee6a68df7dc0daa273c135eef81/static/js/context-menu.js#L48).  I also was a little perturbed by the lack of explicit parsing into an integer, but took what I'll call the "shrug, YOLO JS" approach where it clearly worked and certainly was terse and explicit (and if arrays are going to coerce as needed, why not).

rs=asuth for sure on comprehensive cleanup of this idiom that stops me from cargo culting bad stuff in the future ;)